app/services/invoices/payments/stripe_service.rb

```diff
@@ -18,6 +18,8 @@ def create
           return result
         end
 
+        increment_payment_attempts
```

Comment:
There is something wrong with this approach :/

Since we are updating the payment attempts on every call to this method, and this number is now part of the idempotency key we are sending to stripe (same for gocardless) it basically means we are turning of the idempotency system of the payment provider.

Its already a problem without retry (let's imagine we have a validation error in this method with the `payment.save!`, stripe payment is created but nothing is persisted in our DB) but since we now are adding a retry system open to API/GraphQL it could lead to catastrophic problems (let's imagine someone is calling 15 time the retry end-point at the same time, we could end-up with 15 stripe payments for the same invoice)

I really think we have to put in place some protections before releasing this change:
- Limit the number of retry on a period of time on API/GraphQL
- Ensure a single payment job is enqueued at a time
- Maybe flag processing errors to discriminate them from "code errors" and ensure `provider_payment_id` is always stored somewhere even if the payment object failed to create.

Happy to help you on it if needed ;)

Reply:
Hey @vincent-pochet, thanks for the comment, yeah it would be great to sync regarding it once I'm back :)
However AFAIK we are safe for the scenario you have mentioned. Once we create payment we set `ready_for_payment_processing` flag to `false` (unless we directly receive failed error from payment provider). So, any subsequent retry requests are declined (since this flag is checked before payment retry). We only allow retries once we explicitly receive failed webhook from payment provider (in that case we switch `ready_for_payment_processing` to `true` 
In case in which provider processed request as `succeeded` and if our payment object fails while persisting into DB, we are still safe. Flag `ready_for_payment_processing` will be `false`, and since incoming webhook wouldn't find our payment object in our DB (since creation failed) and also wouldn't find associated invoice so `ready_for_payment_processing` would stay false.

Reply:
Thanks @lovrocolic for these explanation :) (Enjoy your holidays! :p)

Sorry for the noise I didn't saw that part. We are safe on some aspects. We will probably still have to ensure a payment job is enqueued only once at a time and it should be alright. :+1: